pkg/util/converter/prom.go

```diff
@@ -0,0 +1,404 @@
+		pairs = append(pairs, [2]string{k, iter.ReadString()})
+	}
+	return pairs
+}
+
+func readLabelsOrExemplars(iter *jsoniter.Iterator) (*data.Frame, [][2]string) {
+	pairs := make([][2]string, 0, 10)
+	labels := data.Labels{}
+	var frame *data.Frame
+
+	for l1Field := iter.ReadObject(); l1Field != ""; l1Field = iter.ReadObject() {
+		switch l1Field {
+		case "seriesLabels":
+			iter.ReadVal(&labels)
+		case "exemplars":
+			lookup := make(map[string]*data.Field)
+			timeField := data.NewFieldFromFieldType(data.FieldTypeTime, 0)
+			valueField := data.NewFieldFromFieldType(data.FieldTypeFloat64, 0)
+			valueField.Name = labels["__name__"]
+			delete(labels, "__name__")
+			valueField.Labels = labels
+			frame = data.NewFrame("", timeField, valueField)
+			for iter.ReadArray() {
+				for l2Field := iter.ReadObject(); l2Field != ""; l2Field = iter.ReadObject() {
+					switch l2Field {
+					// nolint:goconst
+					case "value":
+						v, _ := strconv.ParseFloat(iter.ReadString(), 64)
+						valueField.Append(v)
+
+					case "timestamp":
+						ts := timeFromFloat(iter.ReadFloat64())
+						timeField.Append(ts)
+
+					case "labels":
+						max := 0
+						for _, pair := range readLabelsAsPairs(iter) {
+							k := pair[0]
+							v := pair[1]
+							f, ok := lookup[k]
+							if !ok {
+								f = data.NewFieldFromFieldType(data.FieldTypeString, 0)
+								f.Name = k
+								lookup[k] = f
+								frame.Fields = append(frame.Fields, f)
+							}
+							f.Append(v)
+							if f.Len() > max {
+								max = f.Len()
+							}
+						}
+						for _, f := range lookup {
+							if f.Len() != max {
+								f.Append("") // no matching label
+							}
+						}
+
+					default:
+						iter.Skip()
+						frame.AppendNotices(data.Notice{
+							Severity: data.NoticeSeverityError,
+							Text:     fmt.Sprintf("unable to parse key: %s in response body", l2Field),
+						})
+					}
+				}
+			}
+		default:
+			v := fmt.Sprintf("%v", iter.Read())
+			pairs = append(pairs, [2]string{l1Field, v})
+		}
+	}
+
+	return frame, pairs
+}
+
+func readMatrixOrVector(iter *jsoniter.Iterator) *backend.DataResponse {
+	rsp := &backend.DataResponse{}
+
+	for iter.ReadArray() {
+		timeField := data.NewFieldFromFieldType(data.FieldTypeTime, 0) // for now!
+		valueField := data.NewFieldFromFieldType(data.FieldTypeFloat64, 0)
+		valueField.Labels = data.Labels{}
+
+		for l1Field := iter.ReadObject(); l1Field != ""; l1Field = iter.ReadObject() {
+			switch l1Field {
+			case "metric":
+				iter.ReadVal(&valueField.Labels)
+
+			case "value":
+				t, v, err := readTimeValuePair(iter)
+				if err == nil {
+					timeField.Append(t)
+					valueField.Append(v)
+				}
+
+			// nolint:goconst
+			case "values":
+				for iter.ReadArray() {
+					t, v, err := readTimeValuePair(iter)
+					if err == nil {
+						timeField.Append(t)
+						valueField.Append(v)
+					}
+				}
+			}
+		}
+
+		valueField.Name = valueField.Labels["__name__"]
+		delete(valueField.Labels, "__name__")
```

Comment:
this is prometheus specific, Loki does not have `__name__`. i would recommend just not doing it, and giving the field a fixed name, like "Value" or "value".  the caller then can customize tha name & labels.

Reply:
🤔 ideally we have the same path for loki+prometheus, I updated it so we set the name to "Value" when the label is missing.  I think this matches expected behavior if a "__name__" label somehow exists

Reply:
1. i see your point. personally i would still prefer to move these kind of customizations to the caller, because i feel like we should do as little "magic" as possible, and considering this is a very easy transform for the user..i'll think more about it :-) 
2. i went to check how this is done currently in Prometheus and Loki, and it is a little different from how this PR does it:
    - Prometheus does not remove the `__name__` label from the field-labels. Loki does remove  the `__name__` from the field-labels, but i THINK this is a bug: 
        - https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/loki/result_transformer.ts#L294 
            - we mutate the object that we get from the caller, and to me this does not seem intentional
            - i think we should change this for grafana9
    - the rest is the same for both Prom & Loki:
        - `field.Name` is always `Value`
        - `field.Config.DisplayNameFromDS` is based on `__name__`
        - `frame.Name` is based on `__name__`